mattermost/desktop
# MM 7970: Tap into system events and idle time to affect user online status based on user activity within the host OS

Before submitting, please confirm you've

Please provide the following information:

**Summary**
Implements  Electron system idle detection ([PowerMonitor](https://electronjs.org/docs/api/power-monitor#powermonitor)) to allow the Desktop app to maintain online status while the app is unfocused, but the user is still active on their computer.

Desktop app maintains an internal inactivity clock with a threshold of 5 minutes, during which updates are sent to the Webapp indicating the user is active. Once the clock exceeds the threshold, updates to the Webapp indicate the user is inactive. System events like screensaver, shutdown, lock, unlock etc. affect status sent to the Webapp.

This PR is part of several related PR's:
Desktop: This PR
Webapp: [PR 2466](https://github.com/mattermost/mattermost-webapp/pull/2466)
Redux: [PR 795](https://github.com/mattermost/mattermost-redux/pull/795)
Server: [PR 10469](https://github.com/mattermost/mattermost-server/pull/10469)

**Issue link**
[MM-7970](https://mattermost.atlassian.net/browse/MM-7970), supported by [MM-14582](https://mattermost.atlassian.net/browse/MM-14582)

**Test Cases**
1. Compile and run the Desktop app.
2. Run a local version of Webapp from PR linked above.
3. Add the local version of the Webapp to the Desktop app.
4. Leave the desktop app visible but in the background for ~10mins while 'using' your computer and observe the status indicator. (It should remain in the online state for the entire time)
5. Leave the desktop app visible but in the background for ~10mins while NOT using your computer and observe the status indicator. (It should go into away mode somewhere around the 6 minute mark).
6. Click on the desktop app to make it active again. Observe the status indicator, it should return to online status.

**Additional Notes**
This is the first phase of a potentially more involved implementation that will include the Webapp properly responding to inactive status caused by system events like shutdown, lock, unlock, screen saver etc. and respond more immediately to status updates.
